crates/sui-framework/sources/governance/genesis.move

```diff
@@ -4,128 +4,206 @@
-            let commission_rate = *vector::borrow(&validator_commission_rates, i);
-            vector::push_back(&mut validators, validator::new(
+            let GenesisValidatorMetadata {
+                name,
+                description,
+                image_url,
+                project_url,
+                sui_address,
+                gas_price,
+                commission_rate,
+                protocol_public_key,
+                proof_of_possession,
+                network_public_key,
+                worker_public_key,
+                network_address,
+                p2p_address,
+                primary_address,
+                worker_address,
+            } = *vector::borrow(&genesis_validators, i);
+
+            let validator = validator::new(
                 sui_address,
-                pubkey,
-                network_pubkey,
-                worker_pubkey,
+                protocol_public_key,
+                network_public_key,
+                worker_public_key,
                 proof_of_possession,
                 name,
                 description,
                 image_url,
                 project_url,
-                net_address,
+                network_address,
                 p2p_address,
                 primary_address,
                 worker_address,
-                // Initialize all validators with uniform stake taken from the subsidy fund.
-                // TODO: change this back to take from subsidy fund instead.
-                option::some(balance::split(&mut sui_supply, initial_validator_stake_mist)),
+                // Initialize all validators with no stake. Stake distribution
+                // will be handled by the 'allocate_tokens' call bellow.
+                option::none(),
                 gas_price,
                 commission_rate,
-                true, // validator is active right away
                 ctx
-            ));
+            );
+
+            vector::push_back(&mut validators, validator);
+
             i = i + 1;
         };
 
+        // Allocate tokens and staking operations
+        allocate_tokens(
+            sui_supply,
+            token_distribution_schedule.allocations,
+            &mut validators,
+            ctx
+        );
+
+        // Activate all validators
+        activate_validators(&mut validators);
+
         sui_system::create(
             validators,
             subsidy_fund,
             storage_fund,
-            governance_start_epoch,
+            genesis_chain_parameters.governance_start_epoch,
             INIT_STAKE_SUBSIDY_AMOUNT,
             protocol_version,
             system_state_version,
-            epoch_start_timestamp_ms,
-            epoch_duration_ms,
+            genesis_chain_parameters.chain_start_timestamp_ms,
+            genesis_chain_parameters.epoch_duration_ms,
             ctx,
         );
 
         clock::create();
+    }
+
+    fun allocate_tokens(
```

Comment:
The code below looks fine to me but I will feel much better if we have some unit tests for it.

Reply:
Checking that 
- the activation_epoch of the stakes is set properly
- the right amount of tokens is sent to everyone
- the next epoch stake, etc fields are set properly since `request_add_stake_at_genesis` has its custom logic different from non-genesis time staking operation..

Reply:
I have a number of checks in rust code, and am planning on adding more. I'm not quite sure the best way to write a move test for this though, due to transferring off the objects.

Reply:
There are some useful functions in `test_scenario` for taking objects out of storage, such as
https://github.com/MystenLabs/sui/blob/370866cb6d4cc9734b91141564600346d07f09a4/crates/sui-framework/sources/test/test_scenario.move#L239-L243
You can have some simple tests that just assign one stake allocation for each staker and use this function to get the `StakedSui` or `Coin<SUI>` and check the amounts. 

And for checking the contents of 0x5 you can use 
https://github.com/MystenLabs/sui/blob/370866cb6d4cc9734b91141564600346d07f09a4/crates/sui-framework/sources/test/test_scenario.move#L342-L346